inc/saivlan.h

```diff
@@ -487,6 +487,170 @@ typedef enum _sai_vlan_stat_t
 
 } sai_vlan_stat_t;
 
+/**
+ * @brief VLAN Stack stage
+ */
+typedef enum _sai_vlan_stack_stage_t
+{
+    SAI_VLAN_STACK_STAGE_INGRESS,
+
+    SAI_VLAN_STACK_STAGE_EGRESS,
+
+} sai_vlan_stack_stage_t;
+
+/**
+ * @brief VLAN Stack action
+ *
+ * If the action is SAI_VLAN_STACK_ACTION_SWAP, the original vlan id of the frame will be replaced
+ * with new vlan id and the new vlan id will be placed in the original tag position.
+ * If the action is SAI_VLAN_STACK_ACTION_PUSH, the original vlan id of the frame will be kept if
+ * exists and the new vlan id will be added as the outer vlan tag.
+ * If the action is SAI_VLAN_STACK_ACTION_POP, the original vlan id of the frame will be removed
+ * if exists.
+ */
+typedef enum _sai_vlan_stack_action_t
+{
+    SAI_VLAN_STACK_ACTION_SWAP,
+
+    SAI_VLAN_STACK_ACTION_PUSH,
+
+    SAI_VLAN_STACK_ACTION_POP,
+
+} sai_vlan_stack_action_t;
+
+/**
+ * @brief Attributes for VLAN Stack
+ */
+typedef enum _sai_vlan_stack_attr_t
+{
+    /**
+     * @brief Start of attributes
+     */
+    SAI_VLAN_STACK_ATTR_START,
+
+    /**
+     * @brief Stage type
+     * If action type is PUSH, this should set to INGRESS
+     * If action type is POP, this should set to EGRESS
+     *
+     * @type sai_vlan_stack_stage_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     */
+    SAI_VLAN_STACK_ATTR_STAGE = SAI_VLAN_STACK_ATTR_START,
+
+    /**
+     * @brief Action type
+     *
+     * @type sai_vlan_stack_action_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     */
+    SAI_VLAN_STACK_ATTR_ACTION,
+
+    /**
+     * @brief Original Vlan ID of the inner or outer
+     * The action will perform on the packets whose vlan tag match the corresponding vlan ID
+     *
+     * @type sai_vlan_stacking_vid_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     */
+    SAI_VLAN_STACK_ATTR_ORIGINAL_VLAN_ID,
+
+    /**
+     * @brief VLAN Stack port object
+     *
+     * @type sai_object_id_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     * @objects SAI_OBJECT_TYPE_PORT, SAI_OBJECT_TYPE_LAG
+     */
+    SAI_VLAN_STACK_ATTR_PORT,
+
+    /**
+     * @brief Applied Vlan ID of the inner or outer
+     * If action type is POP, should not set this attribute
+     *
+     * @type sai_vlan_stacking_vid_t
+     * @flags CREATE_AND_SET
+     * @default 0,0
+     * @validonly SAI_VLAN_STACK_ATTR_ACTION == SAI_VLAN_STACK_ACTION_PUSH or SAI_VLAN_STACK_ATTR_ACTION == SAI_VLAN_STACK_ACTION_SWAP
+     */
+    SAI_VLAN_STACK_ATTR_APPLIED_VLAN_ID,
+
+    /**
+     * @brief The packet priority (3 bits) in the vlan tag
+     * Range from 0 to 7, priority from low to high.
+     * Default 0xFF will inherit the original vlan tag priority
+     *
+     * @type sai_uint8_t
+     * @flags CREATE_AND_SET
+     * @default 0xFF
+     */
+    SAI_VLAN_STACK_ATTR_VLAN_APPLIED_PRI,
```

Comment:
The HLD definition of Priority attribute is not consistent with what has been described here.
"Rule priority from low to high can assign this from 0 to 7. If a packet can match multiple vlan stacking rule, only the highest priority rule will take effect."

could you please give an example, how this attribute is used?

Reply:
I think maybe my statement make mistake easily. In both places, I try to describe the same thing.
0 is the loweset priority and 7 is the highest priority.

For example:
I set a (10-to-20) vlan translation with priority 0 on the port1. After that, I set a (30-to-40) vlan translation with priority 7 on the port1.
Only the (30-to-40) vlan translation will take effect.

If my explanation reasonable, I will then append it into the HLD to make the attribute function more clear.

Reply:
The statement in the HLD is wrong. I will correct it now.
The attribute there is used to fill the "priority" column of the VLAN tag.

If set this value to 0xFF, the priority of the transformed VLAN tag will be equal to the one before the transformation.